content/blog/2025-12-16-stateful-allo.md

```diff
@@ -0,0 +1,115 @@
++++
+title = "Modeling Stateful Accelerators in Allo"
+[extra]
+bio = """
+  Sunwoo Kim is interested in AI hardware. He is pursuing ECE PhD at Cornell.
+"""
+[[extra.authors]]
+name = "Sunwoo Kim"
+link = "https://sunwookim028.github.io/"
++++
+
+# Context and Vision
+Computer architects are building accelerator chips for demanding domains like AI. 
+[Allo](https://cornell-zhang.github.io/allo/) is a Python-embedded domain-specific language to facilitate the design of such accelerators in a modular and composable manner, with users across industry and academia.
+However, the current language doesn't have a support for stateful variables, hence users cannot model practical accelerators that holds intermediate data or results on chip.
+We aimed to add a language feature for users to specify stateful variables with compilation flow to lower them to static variables in Vitis HLS backend.
+
+# Summary of Results
+We successfully added a new type qualifier in Allo to specify stateful variables and compilation flow to lower them into static variables in Vitis HLS, bridged by MLIR.
+We demonstrated programmable stateful accelerator designs for scalar and tensor arithmetics.
+Our most up-to-date code is in this [fork](https://github.com/sunwookim028/allo) and is being merged to the upstream.
+Here's a presentation [video](https://youtu.be/2dKNX0L-iG8?si=Jlyv5rDRoa-c0X9h) for a different course (ECE6775) where this project was also presented.
+
+# Acknowledgement
+This work was done with Joseph Maheshe, Jifeng Wu, Jenny Lee from ECE6775 project team.
+My specific contributions were in suggesting the frontend syntax, implementing the MLIR to HLS backend and writing and running stateful scalar and tensor arithmetic accelerators.
+
+# Design and Implementation
+<img src="./2025-12-16-stateful-allo/overview.png" alt="allo to mlir to hls" width="310"/>
+This is the overview of our compilation flow.
+We'll use this scalar accumulator example to explain each step.
+
+```python
+# A kernel that accumulates values across invocations
+def stateful_kernel(x: int32) -> int32:
+    # 'acc' retains its value between calls
+    acc: stateful(int32) = 0
```

Comment:
It's an interesting choice to use a type qualifier to mark state like this. One thing that's a little weird about it is that this no longer matches Python semantics; if you were to run this code in CPython, the variable would be "fresh" in every invocation of `stateful_kernel`.

A typical Python programmer would accomplish this statefulness like this:

```py
acc = 0
def stateful_kernel(x: int32) -> int32:
    global acc
    acc += x
    return acc
```

I understand that this pattern might not be feasible to recognize in Allo's Python frontend, but it occurred to me to mention as an alternative.